File: Java/1-50/CombinationSum.java
public class CombinationSum {
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        List<List<Integer>> res = new ArrayList<>();
        helper(res, new ArrayList<>(), candidates, 0, target);
        return res;
    }
    void helper(List<List<Integer>> res, List<Integer> curr, int[] candidates, int start, int target) {
        if (target <= 0) {
            if (target == 0) res.add(new ArrayList<>(curr));
            return;
        }
        for (int i = start; i < candidates.length; i++) {
            curr.add(candidates[i]);
            helper(res, curr, candidates, i, target - candidates[i]);
            curr.remove(curr.size() - 1);
        }
    }
}

File: Python/1-50/CombinationSum.py
class Solution(object):
    def combinationSum(self, candidates, target):
        """
        :type candidates: List[int]
        :type target: int
        :rtype: List[List[int]]
        """
        candidates.sort()
        ret = []
        def dfs(start, curr, target):
            if target <= 0:
                if target == 0:
                    ret.append(curr)
                return
            for i in range(start, len(candidates)):
                dfs(i, curr + [candidates[i]], target - candidates[i])
        dfs(0, [], target)
        return ret
